src/all.rs

```diff
@@ -521,33 +522,38 @@ unsafe impl<'a, Trait: ?Sized + TraitQuery> WorldQuery for All<&'a Trait> {
         state: &Self::State,
         access: &mut bevy_ecs::query::FilteredAccess<ComponentId>,
     ) {
+        let mut not_first = false;
+        let mut new_access = access.clone();
         for &component in &*state.components {
             assert!(
                 !access.access().has_write(component),
                 "&{} conflicts with a previous access in this query. Shared access cannot coincide with exclusive access.",
                 std::any::type_name::<Trait>(),
             );
-            access.add_read(component);
-        }
-    }
-
-    #[inline]
-    fn update_archetype_component_access(
-        state: &Self::State,
-        archetype: &bevy_ecs::archetype::Archetype,
-        access: &mut bevy_ecs::query::Access<bevy_ecs::archetype::ArchetypeComponentId>,
-    ) {
-        for &component in &*state.components {
-            if let Some(archetype_component_id) = archetype.get_archetype_component_id(component) {
-                access.add_read(archetype_component_id);
+            if not_first {
+                let mut intermediate = access.clone();
+                intermediate.add_read(component);
+                new_access.append_or(&intermediate);
+                new_access.extend_access(&intermediate);
+            } else {
+                new_access.and_with(component);
+                new_access.access_mut().add_read(component);
+                not_first = true;
```

Comment:
What's the purpose of the `new_access` and `intermediate` variables? Can we not insert directly into the filtered access?

Reply:
This similar to how `AnyOf` is implemented in Bevy. The expected result of the whole method is that `access` contains the result of `(access AND write component1) OR (access AND write component2) OR ...`

The way this is achieved is by computing each `access AND write componentN` in an `intermediate` `FilteredAccess`, and then `OR` them together in an accumulator, which is `new_access` `FilteredAccess`. `new_access` needs to be different from `access` because to compute each intermediate you need the original `access`, so directly modifying `access` will produce incorrect results.

Reply:
Makes sense, thank you for explaining.